File: src/main/java/ru/piko/pikopluginlib/PlayersData/PlayerData.java
package ru.piko.pikopluginlib.PlayersData;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.function.Supplier;

public class PlayerData {
    private final UUID uuid;

    /**
     * A map that stores player-specific data for various plugins and subsystems.
     *
     * <p>The key is a string formatted as "<code>plugin_id.subsystem_id.name</code>", where:</p>
     * <ul>
     *   <li><strong>plugin_id</strong>: The unique identifier for the plugin (e.g., "magic_piko").</li>
     *   <li><strong>subsystem_id</strong>: The identifier for a specific subsystem within the plugin, such as "menu" or "command".</li>
     *   <li><strong>name</strong>: The name or further identifier of the data (e.g., "staff").</li>
     * </ul>
     *
     * <p>Example key: "magic_piko.menu.staff"</p>
     *
     * <p>The value is an instance of {@link APlayerData}, which represents the data associated with that specific plugin and subsystem for the player.</p>
     */
    private final Map<String, APlayerData> playerDataMap;

    public PlayerData(UUID owner) {
        this.uuid = owner;
        this.playerDataMap = new HashMap<>();
    }

    public UUID getUUID() {
        return uuid;
    }

    public boolean hasPlayerOnline() {
        return getOwner() != null;
    }

    public Player getOwner() {
        return Bukkit.getPlayer(uuid);
    }

    public void addData(String id, APlayerData data) {
        playerDataMap.put(id, data);
    }

    public <T extends APlayerData> Optional<T> tryGetData(String id) {
        if (hasData(id)) {
            return Optional.of((T) getData(id));
        }
        return Optional.empty();
    }

    public APlayerData getData(String id) {
        return playerDataMap.get(id);
    }

    public boolean hasData(String id) {
        return playerDataMap.containsKey(id);
    }

    public void removeData(String id) {
        playerDataMap.remove(id);
    }

    /**
     * Retrieves existing data or creates it if absent.
     *
     * @param id The identifier for the data.
     * @param factory The factory method to create the data if it doesn't exist.
     * @param <T> The type of the data.
     * @return The existing or newly created data.
     */
    public <T extends APlayerData> T getOrCreateData(String id, Supplier<T> factory) {
        return (T) playerDataMap.computeIfAbsent(id, k -> factory.get());
    }
}

File: src/main/java/ru/piko/pikopluginlib/PlayersData/PlayerData.kt
package ru.piko.pikopluginlib.PlayersData

import io.netty.buffer.Unpooled
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import ru.piko.pikopluginlib.Api.PikoPluginLibApi
import ru.piko.pikopluginlib.Utils.ByteBufCodec
import java.util.*

class PlayerData(val uuid: UUID) {
	/**
	 * A map that stores player-specific data for various plugins and subsystems.
	 *
	 *
	 * The key is a string formatted as "`author.plugin.name`", where:
	 *
	 *  * **author**: The nickname of the plugin author or the name of the team (e.g., "piko")
	 *  * **plugin**: The unique identifier for the plugin (e.g., "magic").
	 *  * **name**: The name or further identifier of the data (e.g., "staff").
	 *
	 *
	 *
	 * Example key: "piko.magic.staff"
	 *
	 *
	 * The value is an instance of [APlayerData], which represents the data associated with that specific plugin and subsystem for the player.
	 */
	val playerDataMap: MutableMap<String, APlayerData> = HashMap()
	
	val isOnline: Boolean get() = offlineOwner.isOnline
	
	val offlineOwner: OfflinePlayer
		get() = Bukkit.getOfflinePlayer(uuid)
	
	val owner: Player?
		get() = Bukkit.getPlayer(uuid)
	
	fun addData(id: String, data: APlayerData) {
		val reg = PikoPluginLibApi.playerData.registryMap[id] ?: error("PlayerData id: '$id' is not registered")
		playerDataMap[id] = when (reg) {
			is CodecPlayerDataRegistry<*> -> {
				@Suppress("UNCHECKED_CAST")
				val codec = reg.codec as? ByteBufCodec<APlayerData> ?: error("PlayerData id: '$id'. The codec cannot be used.")
				val buffer = Unpooled.buffer()
				codec.encode(buffer, data)
				ByteBufPlayerData(id, buffer)
			}
			is CommonPlayerDataRegistry -> {
				data
			}
			else -> error("Unknown implementation of IPlayerDataRegistry: '${reg::class.java}'")
		}
	}
	
	fun <T : APlayerData?> tryGetData(id: String): Pair<Boolean, T?> {
		if (hasData(id)) {
			val value = getData(id) as? T?
			return Pair(value != null, value)
		}
		return Pair(false, null)
	}
	
	fun getData(id: String): APlayerData? {
		val reg = PikoPluginLibApi.playerData.registryMap[id] ?: error("PlayerData id: '$id' is not registered")
		if (playerDataMap.containsKey(id)) {
			val value = playerDataMap[id]
			if (value != null) when (reg)
			{
				is CodecPlayerDataRegistry<*> -> {
					if (value !is ByteBufPlayerData) return null
					val codec = reg.codec as? ByteBufCodec<APlayerData> ?: return null
					value.buffer.readerIndex(0)
					return codec.decode(value.buffer)
				}
				is CommonPlayerDataRegistry -> {
					return value
				}
				else -> error("Unknown implementation of IPlayerDataRegistry: '${reg::class.java}'")
			}
		}
		return null
	}
	
	fun hasData(id: String): Boolean {
		return playerDataMap.containsKey(id)
	}
	
	fun removeData(id: String) {
		playerDataMap.remove(id)
	}
	
	fun clearStartWith(str: String, ignoreCase: Boolean = false) {
		playerDataMap.keys.filter { it.startsWith(str, ignoreCase) }.forEach { playerDataMap.remove(it)?.also { also -> if (also is ByteBufPlayerData) also.buffer.release() } }
	}
	
	fun clearFunction(function: (Map.Entry<String, APlayerData>) -> Boolean) {
		playerDataMap.entries
			.filter(function)
			.map { it.key }
			.forEach { playerDataMap.remove(it)?.also { also -> if (also is ByteBufPlayerData) also.buffer.release() } }
	}
	
	/**
	 * Retrieves existing data or creates it if absent.
	 *
	 * @param id The identifier for the data.
	 * @param <T> The type of the data.
	 * @return The existing or newly created data.
	</T> */
	inline fun <reified T : APlayerData> getOrCreateData(id: String): T {
//		return playerDataMap.computeIfAbsent(id, function) as T
		val reg = PikoPluginLibApi.playerData.registryMap[id] ?: error("PlayerData id: '$id' is not registered")
		if (playerDataMap.containsKey(id)) {
			val value = playerDataMap[id]
			if (value != null) when (reg)
			{
				is CodecPlayerDataRegistry<*> -> {
					if (value !is ByteBufPlayerData) error("PlayerData id: '$id'. Registration using CodecPlayerDataRegistry involves storing ByteBufPlayerData, which is created and stored automatically, some kind of internal problem has occurred.")
					val codec = reg.codec as? ByteBufCodec<T> ?: error("PlayerData id: '$id'. The codec cannot be used.")
					value.buffer.readerIndex(0)
					return codec.decode(value.buffer)
				}
				is CommonPlayerDataRegistry -> {
					if (value is T) {
						return value as T
					}
				}
				else -> error("Unknown implementation of IPlayerDataRegistry: '${reg::class.java}'")
			}
		}
		val value = reg.create()
		addData(id, value)
		return value as T
	}
}